h2o-algos/src/main/java/hex/optimization/OptimizationUtils.java

```diff
@@ -455,5 +461,131 @@ public GradientInfo ginfo() {
+      // update gradient from constraints transpose(lambda)*h(beta)(value, not changed, active status may change) 
+      // and gram contribution from ck/2*transpose(h(beta))*h(beta)), value not changed but active status may change
+      state.updateConstraintInfo(equalityConstraints, lessThanEqualToConstraints);
+      // calculate new gradient and objective function;
+      _ginfoOriginal =  calGradient(betaCnd, state, gradientSolver, lambdaEqual, lambdaLessThan,
+              equalityConstraints, lessThanEqualToConstraints);
+    }
 
+    public boolean findAlpha(double[] lambdaEqual, double[] lambdaLessThan, ComputationState state,
+                             ConstrainedGLMUtils.LinearConstraints[] equalityConstraints, 
+                             ConstrainedGLMUtils.LinearConstraints[] lessThanEqualToConstraints,
+                             GLM.GLMGradientSolver gradientSolver) {
+      if (_currGradDirIP > 0) {
+        return false;
+      }
+      GLM.GLMGradientInfo newGrad;
+      double[] newCoef;
+      int betaLen = _originalBeta.length;
+      double[] tempDirection = new double[betaLen];
+      boolean firstWolfe;
+      boolean secondWolfe;
+      boolean alphaiChange;
+      for (int index=0; index<_maxIteration; index++) {
+        ArrayUtils.mult(_direction, tempDirection, _alphai);    // tempCoef=alpha_i*direction
+        newCoef = ArrayUtils.add(tempDirection, _originalBeta); // newCoef = coef + alpha_i*direction
+        // calculate constraint values with new coefficients, constraints magnitude square
+        updateConstraintValues(newCoef, Arrays.asList(_coeffNames), equalityConstraints, lessThanEqualToConstraints);
+        calculateConstraintSquare(state, equalityConstraints, lessThanEqualToConstraints);
+        // update gradient from constraints transpose(lambda)*h(beta)(value, not changed, active status may change) 
+        // and gram contribution from ck/2*transpose(h(beta))*h(beta)), value not changed but active status may change
+        state.updateConstraintInfo(equalityConstraints, lessThanEqualToConstraints);
+        // calculate new gradient and objective function;
+        newGrad =  calGradient(newCoef, state, gradientSolver, lambdaEqual, lambdaLessThan,
+                equalityConstraints, lessThanEqualToConstraints);
+        // evaluate if first Wolfe condition is satisfied;
+        firstWolfe = evaluateFirstWolfe(newGrad);
+        // evaluate if second Wolfe condition is satisfied;
+        secondWolfe = evaluateSecondWolfe(newGrad);
+        // return if both conditions are satisfied;
+        if (firstWolfe && secondWolfe) {
+          _newBeta = newCoef;
+          _ginfoOriginal = newGrad;
+          return true;
+        }
+
+        // set alphai if first Wolfe condition is not satisfied, set alpha i if second Wolfe condition is not satisfied;
+        alphaiChange = setAlphai(firstWolfe, secondWolfe);
+        if (!alphaiChange || _alphar < EPS_CS_SQUARE) {
```

Comment:
Even if this condition is true or false, the method returns always false.

Reply:
No, if the first condition is satisfied, it will exit the function and return true:

![image](https://github.com/h2oai/h2o-3/assets/7231712/75962440-2406-4821-8020-6a8eacd21e78)

Reply:
Yes, the first condition makes sense, but if the first condition is false, then even if the second condition is true or false, the method returns false.  So, the second condition does not make sense to me. 

If `if (!alphaiChange || _alphar < EPS_CS_SQUARE)` is true, it returns false on line 592. If the condition is false, it returns false on line 595.  So always false. Am I correct?